File: t4.1-Streams-ej/Grep/Grep.java
import java.io.*;class Grep {      public static void main (String args[]) throws Exception {        if ((args.length < 1) || (3 < args.length))                     throw (new Exception ("Syntax: Grep pattern <dest> <src>"));

        BufferedReader i = new BufferedReader (new InputStreamReader(System.in));        if (args.length == 3) i = new BufferedReader (new FileReader(args[2]));  
        PrintWriter o = new PrintWriter(System.out);        if (args.length > 1) o = new PrintWriter (new FileWriter(args[1]));        String line;        while ((line = i.readLine ()) != null)             {  if (line.matches(args[0])) o.println(line); 
                o.flush(); }        o.close();          i.close();   }}

File: t4.1-Streams-ej/Grep/Grep.rb
# $stdin y $stdout son variables globales donde el sistema almacena
# los valores de acceso estandar STDIN y STDOUT a pantalla y teclado
# Estas variables puden reasignarse a otros dispositivos de
# entrada/salida como ficheros, impresoras, red, etc

if ((ARGV.length < 1) || ( 3 < ARGV.length)) 
	then raise "  Syntax: Grep pattern <src> <dest>  " end

$stdin = File.open(ARGV[1])         if (1 < ARGV.length)
$stdout = File.open(ARGV[2], "w")   if (2 < ARGV.length)
 
while ( !($stdin.eof?) )
	line = $stdin.gets
	$stdout.puts(line) if  (line =~ Regexp.new(ARGV[0]))
end

if ($stdin  != STDIN)  then $stdin.close;  $stdin  = STDIN; end
if ($stdout != STDOUT) then $stdout.close; $stdout = STDOUT; end 
